### PR TITLE
Process async Slack responses in parallel w/thread pool

### PR DESCRIPTION
**Pull request by daveliepmann**
With this PR, we send HTTP callbacks to Slack (in parallel with our direct response) with a queue-moderated fixed thread pool instead of `future`s.

 - the main thread receives requests, putting Slack HTTP callbacks on the thread pool while it sends its own `200 HTTP/OK` response
 - the thread pool keeps a number of worker threads ready to do all the heavy lifting of asking Slack (e.g. for channel or user info), then send responses (e.g. modal screens)

The number of threads in the pool is determined by Goetz's formula using rough estimates of how much time an "average" Slack request has to wait (e.g. on Firebase or Slack API, potentially seconds) and how long we spend processing locally (very little). This ends up being several hundred but should be tuned based on actual use.

---

Prior inconclusive experiments with load testing can be seen as of 49f077c

### CODE REVIEW

File: src/org/sparkboard/server/server.clj [hunk 1]
```diff
@@ -523,14 +543,14 @@
...
-                                                     :slack/team-id ]))
+          (log/info handler-id (select-keys context [:slack/user-id :slack/team-id ]))
           (log/debug :context context)
           (def LAST-CONTEXT context)
-          (binding [slack/*request* context]
-            ((handlers handler-id (fn [& args] (log/error :unhandled-request handler-id args)))
-             (assoc context ::handler-id handler-id)))))
-      (ring.http/ok))))
+          (.submit ^java.util.concurrent.ExecutorService pool
+                   ^Callable #(binding [slack/*request* context]
```
Comment (mhuebert) [inline src/org/sparkboard/server/server.clj:550]: Do we know how error handling differs b/w futures vs ExecutorService? (should we add error-handling/reporting here (like we have in `try-future`)?)

File: src/org/sparkboard/server/server.clj [hunk 2]
```diff
@@ -498,13 +499,32 @@
                               :broadcast/reply-ts (-> payload :message :ts)
                               :broadcast/reply-channel (-> payload :channel :id)}))))}))
 
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;; Async processing (mostly parallel Slack HTTP responses)
+
+(defn goetz
+  "Number of threads to have in a thread pool, per Brian Goetz's 'Java Concurrency in Practice' formula'.
```
Comment (mhuebert) [inline src/org/sparkboard/server/server.clj:507]: 👍

File: src/org/sparkboard/server/server.clj (hunk 1 above)
Comment (daveliepmann) [reply src/org/sparkboard/server/server.clj:550]: Good question, I'll take a look

Comment (daveliepmann) [reply src/org/sparkboard/server/server.clj:550]: Current situation is that exceptions inside the thread pool get *printed*, but do not throw or get logged. With the two commits I just pushed, we now use the snippet from Stuart Sierra to detect and properly *log* those exceptions. It shouldn't be a huge difference.

Note that these commits also change us from `.submit` onto the thread pool to `.execute` on the thread pool. This is a fire-and-forget approach that avoids wrapping the task in a Java Future, which as discussed is unnecessary work.

**Pull request merged**